src/app/components/wrapper-components/part-creation/part-creation.component.spec.ts

```diff
@@ -57,214 +60,278 @@ describe('PartCreationComponent:', () => {
+                // and the candidate has been deselected
                 testUtils.expectElementNotToExist('#selected_firstCandidate');
+                // and the candidate has been removed from the lobby
+                expect(component.currentJoiner).toEqual(JoinerMocks.INITIAL.doc);
+            }));
+            it('should remove candidate from lobby if it directly appears offline', fakeAsync(async() => {
+                spyOn(Utils, 'handleError').and.callFake(() => {});
+                // Given that the page is loaded and there is no candidate yet
+                testUtils.detectChanges();
+                await testUtils.whenStable();
+
+                // When the candidate joins but is directly offline
+                await joueursDAOMock.set('opponent', { ...OPPONENT, state: 'offline' });
+                await mockCandidateArrival();
+                testUtils.detectChanges();
+                await testUtils.whenStable();
+
+                // Then the candidate does not appear on the page
+                testUtils.expectElementNotToExist('#candidate_firstCandidate');
+                // and handleError was called as this is an unexpected situation
+                expect(Utils.handleError).toHaveBeenCalledWith('OnlineGameWrapper: firstCandidate is already offline!');
+            }));
+            it('should not fail if an user has to be removed from the lobby but is not in it', fakeAsync(async() => {
+                spyOn(Utils, 'handleError').and.callFake(() => {});
+                // This could happen if we receive twice the same update to a user that needs to be removed
+
+                // Given a part creation with an candidate
+                testUtils.detectChanges();
+                await testUtils.whenStable();
+                await mockCandidateArrival();
+
+                // If the candidate updates twice with the offline status
+                await joueursDAOMock.update('opponent', { state: 'offline' });
+                testUtils.detectChanges();
+                testUtils.expectElementNotToExist('#candidate_firstCandidate');
+                expect(component.currentJoiner).toEqual(JoinerMocks.INITIAL.doc);
+
+                await joueursDAOMock.update('opponent', { state: 'offline', dummyField: true });
+                testUtils.detectChanges();
+                testUtils.expectElementNotToExist('#candidate_firstCandidate');
+                expect(component.currentJoiner).toEqual(JoinerMocks.INITIAL.doc);
+
+                // Then it is not displayed among the candidates, and no error has been produced
+                testUtils.expectElementNotToExist('#candidate_firstCandidate');
+                expect(Utils.handleError).not.toHaveBeenCalled();
+            }));
+            it('should see candidate reappear if candidates disconnects and reconnects', fakeAsync(async() => {
+                // Given that the page has loaded, a candidate joined and has been chosen as opponent
```

Comment:
given a, non ? Given that ça fait un peu début de démo mathématique :D

Reply:
changé les given that de ce fichier